pkg/resourcemanager/mysql/mysqluser/mysqluser_reconcile.go

```diff
@@ -0,0 +1,319 @@
+			return false, nil
+		}
+
+		// if this is an unmarshall error - igmore and continue, otherwise report error and requeue
+		if _, ok := err.(*json.UnmarshalTypeError); ok {
+			return false, err
+		}
+	}
+
+	adminUser := string(adminSecret["fullyQualifiedUsername"])
+	adminPassword := string(adminSecret[MSecretPasswordKey])
+	fullServerName := string(adminSecret["fullyQualifiedServerName"])
+
+	db, err := s.ConnectToSqlDb(ctx, MDriverName, fullServerName, instance.Spec.DbName, MSqlServerPort, adminUser, adminPassword)
+	if err != nil {
+		instance.Status.Message = errhelp.StripErrorIDs(err)
+		instance.Status.Provisioning = false
+
+		// catch firewall issue - keep cycling until it clears up
+		if strings.Contains(err.Error(), "is not allowed to connect to this MySQL server") {
+			return false, nil
+		}
+
+		// if the database is busy, requeue
+		errorString := err.Error()
+		if strings.Contains(errorString, "Please retry the connection later") {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	// determine our key namespace - if we're persisting to kube, we should use the actual instance namespace.
+	// In keyvault we have to avoid collisions with other secrets so we create a custom namespace with the user's parameters
+	key = GetNamespacedName(instance, mysqlUserSecretClient)
+
+	// create or get new user secret
+	DBSecret := s.GetOrPrepareSecret(ctx, instance, mysqlUserSecretClient)
+	// reset user from secret in case it was loaded
+	user := string(DBSecret[MSecretUsernameKey])
+	if user == "" {
+		user = fmt.Sprintf(requestedUsername)
+		DBSecret[MSecretUsernameKey] = []byte(user)
+	}
+
+	// Publishing the user secret:
+	// We do this first so if the keyvault does not have right permissions we will not proceed to creating the user
+	err = mysqlUserSecretClient.Upsert(
+		ctx,
+		key,
+		DBSecret,
+		secrets.WithOwner(instance),
+		secrets.WithScheme(s.Scheme),
+	)
+	if err != nil {
+		instance.Status.Message = "failed to update secret, err: " + err.Error()
+		return false, err
+	}
+
+	user, err = s.CreateUser(ctx, string(instance.Spec.Server), DBSecret, db)
+	if err != nil {
+		instance.Status.Message = "failed creating user, err: " + err.Error()
+		return false, err
+	}
+
+	// apply roles to user
+	if len(instance.Spec.Roles) == 0 {
+		instance.Status.Message = "No roles specified for user"
+		return false, fmt.Errorf("No roles specified for database user")
+	}
+
+	err = s.GrantUserRoles(ctx, user, string(instance.Spec.Server), instance.Spec.Roles, db)
+	if err != nil {
+		instance.Status.Message = "GrantUserRoles failed"
+		return false, fmt.Errorf("GrantUserRoles failed")
+	}
+
+	instance.Status.Provisioned = true
+	instance.Status.State = "Succeeded"
+	instance.Status.Message = resourcemanager.SuccessMsg
+
+	return true, nil
+}
+
+// Delete deletes a user
+func (s *MySqlUserManager) Delete(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
```

Comment:
There's a massive amount of duplicate boilerplate code in here and `Ensure()`...

Reply:
yeah, there are common operations like GetDB and ConnectToSqlDb etc. However the handlings of the result are not the same during the creation and deletion. At the same time, I found the codes for get admin secret and user secret are exactly the same so could use a common function to return the two secrets. Have created a new issue https://github.com/Azure/azure-service-operator/issues/1146 to optimize. Would like to  provide the basic function for customer testing first.

Reply:
Sounds good, thanks.